nntrainer/layers/grucell.cpp

```diff
@@ -0,0 +1,414 @@
+  wt_idx[GRUCellParams::zrg] =
+    context.requestTensor(zrg_dim, "zrg", Tensor::Initializer::NONE, true,
+                          TensorLifespan::ITERATION_LIFESPAN);
+
+  if (hidden_state_activation_type.get() == ActivationType::ACT_NONE) {
+    hidden_state_activation_type.set(ActivationType::ACT_TANH);
+  }
+  acti_func.setActiFunc(hidden_state_activation_type.get());
+
+  if (recurrent_activation_type.get() == ActivationType::ACT_NONE) {
+    recurrent_activation_type.set(ActivationType::ACT_SIGMOID);
+  }
+  recurrent_acti_func.setActiFunc(recurrent_activation_type.get());
+}
+
+void GRUCellLayer::setProperty(const std::vector<std::string> &values) {
+  auto remain_props = loadProperties(values, grucell_props);
+  LayerImpl::setProperty(remain_props);
+}
+
+void GRUCellLayer::exportTo(Exporter &exporter,
+                            const ExportMethods &method) const {
+  LayerImpl::exportTo(exporter, method);
+  exporter.saveResult(grucell_props, method, this);
+}
+
+void GRUCellLayer::forwarding(RunLayerContext &context, bool training) {
+  const unsigned int unit = std::get<props::Unit>(grucell_props).get();
+  const float dropout_rate = std::get<props::DropOutRate>(grucell_props);
+  const unsigned int max_timestep = std::get<props::MaxTimestep>(grucell_props);
+  const unsigned int timestep = std::get<props::Timestep>(grucell_props);
+
+  Tensor &weight_xh = context.getWeight(wt_idx[GRUCellParams::weight_xh]);
+  Tensor &weight_hh = context.getWeight(wt_idx[GRUCellParams::weight_hh]);
+  Tensor &bias_ih = context.getWeight(wt_idx[GRUCellParams::bias_h]);
+
+  Tensor &input = context.getInput(SINGLE_INOUT_IDX);
+  Tensor &hidden_states =
+    context.getTensor(wt_idx[GRUCellParams::hidden_state]);
+  Tensor &zrg_gates = context.getTensor(wt_idx[GRUCellParams::zrg]);
+  Tensor prev_hidden_state;
+
+  const TensorDim &input_dim = input.getDim();
+  const unsigned int batch_size = input_dim.batch();
+
+  hidden_states.reshape({max_timestep, 1, batch_size, unit});
+  zrg_gates.reshape({max_timestep, 1, batch_size, NUM_GATE * unit});
+
+  Tensor hidden_state = hidden_states.getBatchSlice(timestep, 1);
+  if (!timestep) {
+    prev_hidden_state = Tensor(batch_size, unit);
+    prev_hidden_state.setZero();
+  } else {
+    prev_hidden_state = hidden_states.getBatchSlice(timestep - 1, 1);
+  }
+  Tensor zrg_gate = zrg_gates.getBatchSlice(timestep, 1);
+
+  input.dot(weight_xh, zrg_gate); // x_z, x_r, x_g
+
+  Tensor zr_gate =
+    zrg_gate.getSharedDataTensor({batch_size, 2 * unit}, 0, false);
+  Tensor g_gate =
+    zrg_gate.getSharedDataTensor({batch_size, unit}, 2 * unit, false);
+
+  Tensor weight_hh_zr;
+  Tensor weight_hh_g;
+  weight_hh_zr.copy_with_stride(
+    weight_hh.getSharedDataTensor({1, 1, unit, unit * 2}, 0, false));
+  weight_hh_g.copy_with_stride(
+    weight_hh.getSharedDataTensor({1, 1, unit, unit}, unit * 2, false));
+
+  zr_gate.add_i_strided(prev_hidden_state.dot(weight_hh_zr));
```

Comment:
consider wrapping this statement with `if (!timestep)` than creating a zero valued tensor.

Reply:
Good catch 👍